pkg/expr/nodes.go

```diff
@@ -229,70 +229,89 @@ func (dn *DSNode) Execute(ctx context.Context, now time.Time, _ mathexp.Vars, s
 		Headers: dn.request.Headers,
 	}
 
+	responseType := "unknown"
+	defer func() {
+		if e != nil {
+			responseType = "error"
+		}
+		logger.Debug("Data source queried", "responseType", responseType)
+	}()
+
 	resp, err := s.dataService.QueryData(ctx, req)
 	if err != nil {
 		return mathexp.Results{}, err
 	}
 
 	vals := make([]mathexp.Value, 0)
-	for refID, qr := range resp.Responses {
-		if qr.Error != nil {
-			return mathexp.Results{}, QueryError{RefID: refID, Err: qr.Error}
-		}
-
-		dataSource := dn.datasource.Type
-		if isAllFrameVectors(dataSource, qr.Frames) { // Prometheus Specific Handling
-			vals, err = framesToNumbers(qr.Frames)
-			if err != nil {
-				return mathexp.Results{}, fmt.Errorf("failed to read frames as numbers: %w", err)
+	response, ok := resp.Responses[dn.refID]
+	if !ok {
+		if len(resp.Responses) > 0 {
+			keys := make([]string, 0, len(resp.Responses))
+			for refID := range resp.Responses {
+				keys = append(keys, refID)
 			}
-			return mathexp.Results{Values: vals}, nil
+			logger.Warn("Can't find response by refID. Return nodata", "responseRefIds", keys)
 		}
+		return mathexp.Results{Values: mathexp.Values{mathexp.NoData{}.New()}}, nil
```

Comment:
Apparently this branch revealed a potential bug in test https://github.com/grafana/grafana/blob/4b634fadcfef8a57eb0460c2d488e1c892ed548a/pkg/services/query/query_test.go#L135-L194 

```
expression request error: can only resample type series, got type noData
```

I believe we should fix it

Reply:
This happens because fake plugin returns empty response https://github.com/grafana/grafana/blob/5fb4a7f3b614a9b0f45c11fe9deabff42cfc9bce/pkg/services/query/query_test.go#L478

Reply:
Fixed in https://github.com/grafana/grafana/pull/61708.